### PR TITLE
Using mrtazz's HTTP request library

### PR DESCRIPTION
**Pull request by jedi22**
Importing the HTTP library written by mrtazz.

### CODE REVIEW

File: osquery/http/http.cpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+#include "http.h"
+
+#include <cstring>
+#include <string>
+#include <iostream>
+#include <map>
+
+namespace osquery {
+/** initialize user agent string */
```
Comment (marpaia) [inline osquery/http/http.cpp]: + newline above this

File: osquery/http/http.cpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+const char* RestClient::user_agent = "osquery/1.4.4";
+/** initialize authentication variable */
```
Comment (marpaia) [inline osquery/http/http.cpp]: newline above this

File: osquery/http/http.cpp [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+std::string RestClient::user_pass = std::string();
+/** Authentication Methods implementation */
```
Comment (marpaia) [inline osquery/http/http.cpp]: newline above this

File: osquery/http/http.cpp [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+void RestClient::clearAuth() { RestClient::user_pass.clear(); }
+void RestClient::setAuth(const std::string& user, const std::string& password) {
```
Comment (marpaia) [inline osquery/http/http.cpp]: newline above this

File: osquery/http/http.cpp [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+  /** create return struct */
+  RestClient::response ret = {};
+
+  // use libcurl
+  CURL* curl = NULL;
+  CURLcode res = CURLE_OK;
+
+  curl = curl_easy_init();
+  if (curl) {
+    /** set basic authentication if present*/
```
Comment (marpaia) [inline osquery/http/http.cpp]: // instead of /***/

File: osquery/http/http.cpp [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+    if (RestClient::user_pass.length() > 0) {
+      curl_easy_setopt(curl, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+      curl_easy_setopt(curl, CURLOPT_USERPWD, RestClient::user_pass.c_str());
+    }
+    /** set user agent */
```
Comment (marpaia) [inline osquery/http/http.cpp]: same here and all of the other in-function comments

File: osquery/http/http.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+#include <map>
+#include <cstdlib>
+#include <algorithm>
+
+namespace osquery {
+
+class RestClient {
+ public:
+  /**
+   * public data definitions
```
Comment (marpaia) [inline osquery/http/http.h]: @brief

File: osquery/http/http.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+   */
+  typedef std::map<std::string, std::string> headermap;
+
+  /** response struct for queries */
+  typedef struct {
```
Comment (marpaia) [inline osquery/http/http.h]: /// instead of /****/

File: osquery/http/http.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    int code;
+    std::string body;
+    headermap headers;
+  } response;
+  /** struct used for uploading data */
```
Comment (marpaia) [inline osquery/http/http.h]: newline above this and /// instead of /** */

File: osquery/http/http.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  typedef struct {
+    const char* data;
+    size_t length;
+  } upload_object;
+
+  /** public methods */
```
Comment (marpaia) [inline osquery/http/http.h]: rm

Comment (marpaia) [inline osquery/http/http.h]: maybe change this to `public:`? are the other things above this NOT public?

File: osquery/http/http.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  // Auth
+  static void clearAuth();
+  static void setAuth(const std::string& user, const std::string& password);
```
Comment (marpaia) [inline osquery/http/http.h]: newlines and comments

**Comment by theopolis** [comment]
I don't think we should convert any code we drop in. If we don't need to modify the CFG or code at all we should consider moving this into third-party?

If we are going to heavily edit the code, then perhaps we can codmod the comments and apply our clang-formatting rules.

**Comment by marpaia** [comment]
This code is clang-formatted. We need to make sure the license is "included in all copies or substantial portions of the Software." Updating comments and style to fit with the rest of osquery is fine IMO. If we make more significantly dramatic changes, we can remove the original license.

**Comment by theopolis** [comment]
I don't mind so much about the license as I do codemod-ing a drop in implementation.

File: osquery/http/http.cpp (hunk 2 above)
Comment (theopolis) [inline osquery/http/http.cpp]: Use https://github.com/facebook/osquery/blob/master/include/osquery/core.h#L40 for the version

**Pull request closed without merging**